src/Servers/Kestrel/shared/KnownHeaders.cs

```diff
@@ -647,27 +651,43 @@ protected override bool TryGetValueFast(string key, out StringValues value)
             switch (key.Length)
             {{{Each(loop.HeadersByLength, byLength => $@"
                 case {byLength.Key}:
-                    {{{Each(byLength, header => $@"
-                        if (""{header.Name}"".Equals(key, StringComparison.OrdinalIgnoreCase))
-                        {{{(header.Identifier == "ContentLength" ? @"
-                            if (_contentLength.HasValue)
-                            {
-                                value = HeaderUtilities.FormatNonNegativeInt64(_contentLength.Value);
-                                return true;
-                            }
-                            return false;" : $@"
-                            if ({header.TestBit()})
-                            {{
-                                value = _headers._{header.Identifier};
-                                return true;
-                            }}
-                            return false;")}
-                        }}")}
-                    }}
-                    break;")}
+                {{{Each(byLength.OrderBy(h => !h.PrimaryHeader), header => $@"
+                    if (ReferenceEquals(HeaderNames.{header.Identifier}, key))
```

Comment:
Is manually calling object.ReferenceEquals and then following up with string.Equals StringComparison.OrdinalIgnoreCase really faster than doing the string comparison up front? If so, shouldn't we just add an object.ReferenceEquals fast path to the start of string.Equals? Or at the very least add a helper method to Kestrel that does this?

Also, AFIACT the body of this if block is identical to the body of the string.Equals if block below. Wouldn't it be better to just add `|| HeaderNames.{header.Identifier}.Equals(key, StringComparison.OrdinalIgnoreCase)` to this if condition even if it isn't part of some helper method?

Reply:
string.Equals does have a ReferenceEquals at the start (though its uninlinable since its kinda chunky)

The two issues this resolves are
1. Quick inlined reference check
2. Check all the references first before checking IgnoreOrdinalCase

So if there were 3 headers of that length; and it was the 3rd header it would do

ReferenceEquals, IgnoreOrdinalCase, ReferenceEquals, IgnoreOrdinalCase, ReferenceEquals => match

With this approach it instead does

ReferenceEquals, ReferenceEquals, ReferenceEquals => match

Only if none of the ReferenceEquals match does it then start doing IgnoreOrdinalCase. 

Noticed this issue after adding `TraceParent` and trying to workout why `ContentMD5` was being tested IgnoreOrdinalCase.